### ISSUE TITLE
Add support for recursive ascent parser generation

### ISSUE THREAD

**Issue by RReverser** [DmitrySoshnikov/syntax#51]
It looks like currently the tool just generates table-driven parsers for various languages, which is often suboptimal.

Recursive ascent parsers are fully compatible with (LA)LR grammars, but they instead use native language features (functions, switches, etc.) so that they can be statically analysed and, as a result, much better optimised by the target compiler or JIT.

It would be great if `syntax` could generate such parsers for various langs.

**Comment by DmitrySoshnikov** [comment]
Yeah, great idea; just recently I was thinking to add support for recursive descent (from LL(1) grammars), but the recursive ascent sounds event better. I also agree that it's better for static analysis, etc. I'll need to find time for it, and of course will appreciate a PR; this will be simply awesome addition to the tool!

For recursive descent, potentially can accept left-recursive grammars as well (need to implement left-recursion eliminator, #50).
